### PR TITLE
macos: Initial Support

### PR DESCRIPTION
**Pull request by SamoZ256**
This is the updated version of the MacOS backend. It shouldn't interfere with the linux backend now.

### CODE REVIEW

**Comment by SamoZ256** [comment]
I also had to increase the size of the `formats[32]` array to 64, since I was getting `VK_INCOMPLETE` from `vkGetPhysicalDeviceSurfaceFormatsKHR`

File: src/macos/common_priv.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    if (value < 0) {
+        return 0;
+    }
+    if (value > UINT16_MAX) {
+        return UINT16_MAX;
+    }
+    return (uint16_t)value;
+}
+
+enum wsi_api {
```
Comment (markbolhuis) [inline src/macos/common_priv.h:18]: Unused?

File: src/macos/utils.c
Comment (markbolhuis) [inline src/macos/utils.c:1]: This and utils.h don't need to be included right now. wsi_list is unused.

Does macOS come with any native library providing a generic list?

Comment (SamoZ256) [reply src/macos/utils.c:1]: Yes, it’s called NSMutableArray

**Comment by markbolhuis** [comment]
Can you change the commit string to something like: `macos: Initial Support`

I'd like all commits to be prefixed with the backend affected if appropriate.

File: src/meson.build [hunk 2]
```diff
@@ -1,3 +1,6 @@
 if build_machine.system() == 'linux'
   subdir('linux')
 endif
+if build_machine.system() == 'darwin'
```
Comment (markbolhuis) [inline src/meson.build:4]: Use `elif`

https://mesonbuild.com/Syntax.html#if-statements

File: src/macos/common_priv.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#ifndef WSI_SRC_COCOA_COMMON_PRIVATE_H
+#define WSI_SRC_COCOA_COMMON_PRIVATE_H
+
+#include "wsi/common.h"
+
+static inline uint16_t
+wsi_xcb_clamp(int32_t value)
```
[the remark this answers is not in the document]
Comment (SamoZ256) [reply src/macos/common_priv.h:7]: MacOS uses NSRect which contains CGFloat, but that't actually double on most platforms

File: examples/vkcube/vkcube.c [hunk 4]
```diff
@@ -1306,6 +1306,9 @@ demo_create_instance(struct demo *demo)
     VkInstanceCreateInfo instance_info = {
         .sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
         .pNext = NULL,
+#ifdef __APPLE__
```
Comment (markbolhuis) [inline examples/vkcube/vkcube.c:1309]: I'd prefer a check to see if `inst_exts` contains "VK_KHR_portability_enumeration", and if so set the flag.

File: src/macos/utils.c
Comment (SamoZ256) [reply src/macos/utils.c:1]: Yes, I can do that in a minute

File: src/macos/common_priv.h (hunk 1 above)
Comment (SamoZ256) [reply src/macos/common_priv.h]: Ok removed

**Comment by SamoZ256** [comment]
I think I am finished. About the NSMutableArray, where am I supposed to use it? Because I do not think there's a place where I use mutable lists

**Comment by markbolhuis** [comment]
> About the NSMutableArray, where am I supposed to use it?

I meant in general, and not right now. That if you need to use a list in the future that you should use any native type first.

**Comment by markbolhuis** [comment]
Looks OK to me. Please squash into a single commit. 
Also make sure the first line of the commit message is formatted as asked here: https://github.com/markbolhuis/libwsi/pull/3#issuecomment-1622049703

**Comment by SamoZ256** [comment]
Finally, I got it. Git gives me headaches :)

**Pull request merged**